src/curve-fed/CurveFed.sol

```diff
@@ -0,0 +1,226 @@
+    }
+
+    /**
+    * @notice Set limit of the max loss setable by guardian.
+    * @param newGuardianMaxLossBpsLimit The maximum loss allowed by basis points 1 = 0.01%
+    */
+    function setGuardianMaxLossBpsLimit(uint newGuardianMaxLossBpsLimit) external onlyRole(gov){
+        require(newGuardianMaxLossBpsLimit <= 10000, "Max loss > 100%");
+        guardianMaxLossBpsLimit = newGuardianMaxLossBpsLimit;   
+    }
+    
+    /**
+    * @notice Set limit of the supply ceiling
+    * @param newSupplyCeiling The new dola supply ceiling of the fed
+    */ 
+    function setSupplyCeiling(uint newSupplyCeiling) external onlyRole(gov) {
+        supplyCeiling = newSupplyCeiling;
+    }
+
+    /**
+    * @notice Deposits amount of dola tokens into convex
+    * @param amount Amount of dola token to deposit into convex
+    */
+    function expansion(uint amount) external onlyRole(chair){
+        require(amount + dolaSupply <= supplyCeiling, "Expansion above ceiling");
+        dolaSupply += amount;
+        dola.mint(address(this), amount);
+        metapoolDeposit(amount, maxLossExpansionBps);
+        emit Expansion(amount);
+    }
+
+    /**
+    * @notice Withdraws an amount of dola token to be burnt, contracting DOLA dolaSupply
+    * @dev Be careful when setting maxLoss parameter. There will almost always be some slippage when trading.
+    * For example, slippage + trading fees may be incurred when withdrawing from a Curve pool.
+    * On the other hand, setting the maxLoss too high, may cause you to be front run by MEV
+    * sandwhich bots, making sure your entire maxLoss is incurred.
+    * Recommended to always broadcast withdrawl transactions(contraction & takeProfits)
+    * through a frontrun protected RPC like Flashbots RPC.
+    * @param amountDola The amount of dola tokens to withdraw. Note that more tokens may
+    be withdrawn than requested.
+    */
+    function contraction(uint amountDola) external onlyRole(chair){
+        //Calculate how many lp tokens are needed to withdraw the dola
+        uint crvLpNeeded = lpForDola(amountDola);
+        require(crvLpNeeded <= crvLpSupply(), "Not enough crvLP tokens");
+
+        //Withdraw DOLA from curve pool
+        uint dolaWithdrawn = metapoolWithdraw(amountDola, maxLossWithdrawBps);
+        require(dolaWithdrawn > 0, "Must contract");
+        uint burnAmount = _burnAndPay();
+        emit Contraction(burnAmount);
+    }
+
+    /**
+    * @notice Withdraws every remaining crvLP token. Can take up to maxLossWithdrawBps in loss, compared to dolaSupply.
+    * It will still be necessary to call takeProfit to withdraw any potential rewards.
+    */
+    function contractAll() external onlyRole(chair){
+        uint dolaMinOut = dolaSupply * (10_000 - maxLossWithdrawBps) / 10_000;
+        crvMetapool.remove_liquidity_one_coin(crvLpSupply(), 0, dolaMinOut);
+        uint burnAmount = _burnAndPay();
+        emit Contraction(burnAmount);
+    }
```

Comment:
This is vulnerable to massive sandwich attacks. If profit accrues as `virtual_price`, the difference between `dolaSupply` and `virtual_price` of LP tokens will grow over time. A fed that has made 5% profit on it's `dolaSupply` will be sandwichable for 5% + `maxLossWithdrawBps` - slippage, assuming DOLA is at parity.
```suggestion
    function contractAll() external onlyRole(chair){
        uint dolaMinOut = crvLpSupply() * crvMetapool.get_virtual_price() * (10_000 - maxLossWithdrawBps) / 10_000;
        crvMetapool.remove_liquidity_one_coin(crvLpSupply(), 0, dolaMinOut);
        uint burnAmount = _burnAndPay();
        emit Contraction(burnAmount);
    }
```
Is a better solution.